File: 15/solution.py
import sys


def solve(line: str, until: int):
    current_digit = 0
    digits_last_seen = {}
    start_digit_list = [int(n) for n in line.split(',')]
    for i in range(len(start_digit_list)):
        digits_last_seen[start_digit_list[i]] = i
        current_digit = start_digit_list[i]
    for i in range(len(start_digit_list)-1, until-1):
        age = 0
        if current_digit in digits_last_seen:
            age = i - digits_last_seen[current_digit]
        digits_last_seen[current_digit] = i
        current_digit = age
    return current_digit


def part1(lines):
    return solve(lines[0][:-1], 2020)


def part2(lines):
    return solve(lines[0][:-1], 30000000)


try:
    f = open(sys.argv[1], "r")
except IOError:
    print("Error opening the file, try again")
    sys.exit(1)
with f:
    rawLines = f.readlines()
    f.close()
    # digitArray = [int(s) for s in lines]
    print(
        f"Part 1 answer:{part1(rawLines)} Part 2 answer: {part2(rawLines)}")

File: 15/solution.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

func solve(digitString string, until int) int {
	currentDigit := 0
	digitsLastSeen := make(map[int]int)
	for index, digit := range strings.Split(digitString, ",") {
		val, err := strconv.Atoi(digit)
		if err != nil {
			log.Fatal(err)
		}
		digitsLastSeen[val] = index
		currentDigit = val
	}
	startingDigitsCount := len(digitsLastSeen)
	for i := startingDigitsCount - 1; i < until-1; i++ {
		age := 0
		val, exists := digitsLastSeen[currentDigit]
		if exists {
			age = i - val
		}
		digitsLastSeen[currentDigit] = i
		currentDigit = age
	}
	return currentDigit
}

func part1(lines []string) int {
	return solve(lines[0], 2020)
}

func part2(lines []string) int {
	return solve(lines[0], 30000000)
}

func main() {
	flag.Parse()
	if len(flag.Args()) == 0 {
		log.Fatal("Provide the path to input")
	}
	file, err := os.Open(flag.Args()[0])
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)

	var lines []string
	for scanner.Scan() {
		line := scanner.Text()
		lines = append(lines, line)
	}

	fmt.Printf("Part 1 answer:%v\nPart 2 answer:%v\n", part1(lines), part2(lines))
}
